dbt/adapters/base/connections.py

```diff
@@ -0,0 +1,317 @@
+            # or the connection has been closed, put it back on the available
+            # list
+            with self.lock:
+                self.available.append(to_release)
+
+    def cleanup_all(self):
+        with self.lock:
+            for name, connection in self.in_use.items():
+                if connection.state != 'closed':
+                    logger.debug("Connection '{}' was left open."
+                                 .format(name))
+                else:
+                    logger.debug("Connection '{}' was properly closed."
+                                 .format(name))
+
+            conns_in_use = list(self.in_use.values())
+            for conn in conns_in_use + self.available:
+                self.close(conn)
+
+            # garbage collect these connections
+            self.in_use.clear()
+            self._set_initial_connections()
+
+    @abc.abstractmethod
+    def begin(self, name):
+        """Begin a transaction. (passable)
+
+        :param str name: The name of the connection to use.
+        """
+        raise dbt.exceptions.NotImplementedException(
+            '`begin` is not implemented for this adapter!'
+        )
+
+    def get_if_exists(self, name):
+        if name is None:
+            name = 'master'
+
+        if self.in_use.get(name) is None:
+            return
+
+        return self.get(name)
+
+    @abc.abstractmethod
+    def commit(self, connection):
+        """Commit a transaction. (passable)
+
+        :param str name: The name of the connection to use.
+        """
+        raise dbt.exceptions.NotImplementedException(
+            '`commit` is not implemented for this adapter!'
+        )
+
+    def _rollback_handle(self, connection):
+        """Perform the actual rollback operation."""
+        connection.handle.rollback()
+
+    def _rollback(self, connection):
+        """Roll back the given connection.
+
+        The connection does not have to be in in_use or available, so this
+        operation does not require the lock.
+        """
+        if dbt.flags.STRICT_MODE:
+            Connection(**connection)
+
+        if connection.transaction_open is False:
+            raise dbt.exceptions.InternalException(
+                'Tried to rollback transaction on connection "{}", but '
+                'it does not have one open!'.format(connection.name))
+
+        logger.debug('On {}: ROLLBACK'.format(connection.name))
+        self._rollback_handle(connection)
+
+        connection.transaction_open = False
+
+        return connection
+
+    @classmethod
+    def close(cls, connection):
+        if dbt.flags.STRICT_MODE:
+            assert isinstance(connection, Connection)
+
+        # if the connection is in closed or init, there's nothing to do
+        if connection.state in {'closed', 'init'}:
+            return connection
+
+        # On windows, sometimes connection handles don't have a close() attr.
+        if hasattr(connection.handle, 'close'):
```

Comment:
it seems weird to me that we wouldn't call `close()` on the handle on Windows. Is this right?

Reply:
This is actually not new to my code! I saw the `hasattr` check, thought "that's silly and can't be right", removed it, tested locally and found everything fine, and then appveyor failed.

So yeah, sometimes on windows the handle might not have a `close`, a fact I've determined empirically.

Reply:
i am at once shocked and not surprised :)